index.js

```diff
@@ -155,31 +160,50 @@ class AbstractBtpPlugin extends EventEmitter {
         data: Buffer.from(secret, 'utf8')
       }]
 
-      await new Promise((resolve, reject) => {
-        this._ws.on('open', async () => {
-          debug('connected to server')
-
-          await this._call(null, {
-            type: BtpPacket.TYPE_MESSAGE,
-            requestId: await _requestId(),
-            data: { protocolData }
-          })
-
-          resolve()
+      this._ws.on('open', async () => {
+        debug('connected to server')
+        await this._call(null, {
+          type: BtpPacket.TYPE_MESSAGE,
+          requestId: await _requestId(),
+          data: { protocolData }
         })
-
-        this._ws.on('message', this._handleIncomingWsMessage.bind(this, this._ws))
+        this.emit('connect')
       })
+
+      this._ws.on('message', this._handleIncomingWsMessage.bind(this, this._ws))
+      await this._ws.open('ws://' + host) // TODO: wss
     }
 
     if (this._connect) {
       await this._connect()
     }
 
+    await new Promise((resolve, reject) => {
+      this.once('connect', resolve)
+      this.once('disconnect', reject)
+    })
     this._connected = true
   }
 
+  async _closeIncomingSocket (socket) {
+    socket.removeAllListeners()
+    socket.once('message', () => {
+      try {
+        ws.send(BtpPacket.serializeError({
+          code: 'F00',
+          name: 'NotAcceptedError',
+          data: 'This connection has been ended',
```

Comment:
Might be good to explain *why* it ended, e.g. "Connection closed because this user has opened another connection in the meantime." If there are different reasons, just pass the message as a parameter.